dvc/remote/gs.py

```diff
@@ -20,35 +23,47 @@ def dynamic_chunk_size(func):
     @wraps(func)
     def wrapper(*args, **kwargs):
         import requests
-        from google.cloud.storage.blob import Blob, _DEFAULT_CHUNKSIZE
+        from google.cloud.storage.blob import _DEFAULT_CHUNKSIZE
 
-        # Default chunk size for gs is 100M, which might be too much for
-        # particular network (see [1]). So if we are getting ConnectionError,
-        # we should try lowering the chunk size until we reach the minimum
-        # allowed chunk size of 256K. Also note that `chunk_size` must be a
-        # multiple of 256K per the API specification.
+        # `ConnectionError` may be due to too large `chunk_size`
+        # (see [#2572]) so try halving on error.
+        # Note: start with 40 * [default: 256K] = 10M.
+        # Note: must be multiple of 256K.
         #
-        # [1] https://github.com/iterative/dvc/issues/2572
+        # [#2572]: https://github.com/iterative/dvc/issues/2572
 
         # skipcq: PYL-W0212
-        multiplier = int(_DEFAULT_CHUNKSIZE / Blob._CHUNK_SIZE_MULTIPLE)
+        multiplier = 40
         while True:
             try:
                 # skipcq: PYL-W0212
-                chunk_size = Blob._CHUNK_SIZE_MULTIPLE * multiplier
+                chunk_size = _DEFAULT_CHUNKSIZE * multiplier
                 return func(*args, chunk_size=chunk_size, **kwargs)
             except requests.exceptions.ConnectionError:
-                multiplier = int(multiplier / 2)
+                multiplier //= 2
                 if not multiplier:
                     raise
 
     return wrapper
 
 
 @dynamic_chunk_size
-def _upload_to_bucket(bucket, from_file, to_info, **kwargs):
-    blob = bucket.blob(to_info.path, **kwargs)
-    blob.upload_from_filename(from_file)
+def _upload_to_bucket(bucket, from_file, to_info, chunk_size=None, **kwargs):
+    blob = bucket.blob(to_info.path, chunk_size=chunk_size, **kwargs)
+    with Tqdm(
+        desc=to_info.path, total=os.path.getsize(from_file), bytes=True
+    ) as pbar:
+        with io.open(from_file, mode="rb") as fd:
+            raw_read = fd.read
+
+            def read(size=chunk_size):
+                res = raw_read(size)
+                if res:
+                    pbar.update(len(res))
+                return res
+
+            fd.read = read
```

Comment:
yes I thought of wrapping but opted to patch like this instead since

1. It's easier to see what's happening/debug
2. It's shorter/requires less code to implement
3. I didn't think it would be reused.

There are a few minor issues I have with the GDrive PR but I'd rather both this PR and the GDrive one are merged as is and then I can unify both of them nicely using wrappers.

Reply:
actually the wrapping of file-like objects is a nice feature request from https://github.com/tqdm/tqdm/issues/84 which dropped off my radar

Reply:
Feel free to give feedback at https://github.com/tqdm/tqdm/pull/844 (going to make this easier).

Reply:
@casperdcl, so probably after https://github.com/tqdm/tqdm/pull/844 gets merged, we should create a ticket to wrapattrs here, right?

Reply:
The issue with this code is that it can't be reused without copy paste. And that is desirable.

Reply:
yes, of course, but wrapping things to make them reusable without knowing that they actually need to be reused is a waste. In this case it turns out there is indeed another place for use - but it's in a PR in tandem (#2551) so it's best to wait for both PRs to be merged and then fix later. Past that there's the upstream tqdm/tqdm#844 which will fix this.

Reply:
Wrapping things for abstraction is not waste, now a function does two things uploading/downloading and some low-level stuff, which should be abstracted at some point. So it's not only about reuse.

I agree that this can wait. So my resolution:
- create a ticket about wrapping this, 
- mention both GS and GDrive code there, 
- resolve this conversation

Reply:
created #2829